File: src/main/scala/scala_spark/spark_book/第6章/Accumulator.scala
package scala_spark.spark_book.第6章

import org.apache.spark.{SparkConf, SparkContext}


//Accumulator变量主要用于多个节点对一个变量进行共享性操作（累加操作即多个任务对一个变量并行操作的功能
// 但是task只能对Accumulator执行累加操作，不能读取它的值，只有Driver程序可以读取Accumulator的值）

object Accumulator {
  def main(args: Array[String]): Unit = {
    val conf=new SparkConf().setAppName("accumulator").setMaster("local")
    val sc=new SparkContext(conf)
    val sumaccumulator=sc.accumulator(0)
    val arr=Array(1,2,3,4,5)
    val inputRDD=sc.parallelize(arr)
    inputRDD.foreach(num=>sumaccumulator+=num)
    println(sumaccumulator.value)
  }
}

File: src/main/java/java_spark/spark_book/第6章/accumulator.java
package java_spark.spark_book.第6章;

import org.apache.spark.Accumulator;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.VoidFunction;

import java.util.Arrays;
import java.util.List;
//累加器用来对信息进行聚合即将工作中的节点的值聚合到驱动器程序中的简单语法
//工作节点中的任务不能访问累加器中的值，对于工作节点中的任务，累加器是一个只写变量，只能由驱动程序才能访问
public class accumulator {
    public static void main(String[] args) {
        SparkConf conf=new SparkConf().setAppName("broadcast_var").setMaster("local");
        JavaSparkContext sc=new JavaSparkContext(conf);
        final Accumulator<Integer> SumAccumulator=sc.accumulator(0);
        List<Integer> list= Arrays.asList(1,2,3,4);
        JavaRDD<Integer>InputRDD=sc.parallelize(list);
        InputRDD.foreach(new VoidFunction<Integer>() {
            public void call(Integer integer) throws Exception {
                SumAccumulator.add(integer);
            }
        });
        System.out.println(SumAccumulator.value());
    }
}
